reference/SDK: rework SDK docs,cover container SDK (#195)
flatcar-linux/flatcar-docs: docs/reference/developer-guides/sdk-building-production-images.md

```
  main
  ...
   +-- Alpha-2983.0.0
   |        +-- Beta-2983.1.0
   |                +--- Stable-2983.2.0
   |                +--- Stable-2983.2.1
   |
   +-- Alpha-3005.0.0
   +-- Alpha-3005.0.1
   |
   +-- Alpha-3033.0.0
   |        +-- Beta-3033.1.0
   |        +-- Beta-3033.1.1
   |                +--- Stable-3033.2.0
   |
   +-- Alpha-3046.0.0
   |
   +-- Alpha-3066.0.0
   |        +-- Beta-3066.1.0
  ...
```

## On versioning

For Flatcar versioning, the scripts repo is authoritative: 
Versioning is controlled by the [`version.txt` file in the scripts repo](https://github.com/flatcar-linux/scripts/blob/main/sdk_container/.repo/manifests/version.txt) as well as via the commit pointers of the script repo's `portage-stable` and `coreos-overlay` gitmodules.
`version.txt` contains version strings for both the SDK version as well as the OS image version.

Core idea is that a simple
```shell
git checkout 3033.2.0
git submodules update --init --reqursive
```
will set up the scripts repo for development on top of Flatcar release `3033.2.0`.

> Keeping `version.txt`,and submodules in sync, updating version strings, and generating version tags is one of the main concerns of the build automation scripts.

Reviewer: What commands should be used to set the new submodule references? I want to add them to https://github.com/flatcar-linux/scripts/blob/main/tag_release to make sure we update them from now on even if we didn't make the switch.

Author: There are library functions for this; should be straightforward. We need 4 or 5 inputs depending on what we are releasing. Please note that the functions only care about the `scripts` repo so they expect either commit IDs or tags that already exist in these repos. Furthermore, the scripts only handle tagging, not branching. It is expected that the `scripts` repo is already on the correct branch.

For new major releases:
1. coreos-overlay commit-ish, tag, or branch
2. portage-stable commit-ish, tag, or branch
3. new SDK version number (to be used for both SDK and OS)
4. new release version tag name (channel + version number)

For channel propagation and other incremental releases that do not introduce a new major version we additionally need
4. new OS image version number

In the scripts repo, run
```shell
source sdk_lib/sdk_container_common.sh
source ci-automation/ci_automation_common.sh

update_submodules <coreos-overlay commit-ish> <portage-stable commit-ish>

create_versionfile <SDK-version> [<os-version>]

update_and_push_version <release-tag>
```

e.g. for a new Alpha-3110 based on the current main branches:
```shell
source sdk_lib/sdk_container_common.sh
source ci-automation/ci_automation_common.sh

update_submodules main main
create_versionfile 3110.0.0
update_and_push_version alpha-3110.0.0
```

Or for a Beta-3110.1.0 based on the above Alpha:
```shell
source sdk_lib/sdk_container_common.sh
source ci-automation/ci_automation_common.sh

update_submodules flatcar-3110 flatcar-3110
create_versionfile 3110.1.0
update_and_push_version beta-3110.1.0
```

Reviewer: This is not really what we do for the release process, a branch gets created for a new major release first, we don't use `main` to tag the Alpha.
I would say we should not just push a lone tag but actually have it be a commit first on the release maintenance branch.

Reviewer: I've filed a PR now for our release tooling: https://github.com/flatcar-linux/flatcar-build-scripts/pull/91

> Versioning is controlled by the [`version.txt` file in the scripts repo](https://github.com/flatcar-linux/scripts/blob/main/sdk_container/.repo/manifests/version.txt) as well as via the commit pointers of the script repo's `portage-stable` and `coreos-overlay` gitmodules.

Author: Absolutely. I just left it there to get a first iteration of the containerised SDK / versioned git submodules out. One of the design goals of that first iteration was to reproduce *exactly* what users are used to from the `cork` SDK, which included putting the versionfile to a well-known place.
We can absolutely move it elsewhere in a follow-up operation; I feel there's a lot of room for more simplifications after we have established the groundwork.

> Versioning is controlled by the [`version.txt` file in the scripts repo](https://github.com/flatcar-linux/scripts/blob/main/sdk_container/.repo/manifests/version.txt) as well as via the commit pointers of the script repo's `portage-stable` and `coreos-overlay` gitmodules.

Reviewer: Maybe a symlink from `/version.txt` could help? For now people will notice the file when it gets modified automatically and shows up red in `git status`

> Versioning is controlled by the [`version.txt` file in the scripts repo](https://github.com/flatcar-linux/scripts/blob/main/sdk_container/.repo/manifests/version.txt) as well as via the commit pointers of the script repo's `portage-stable` and `coreos-overlay` gitmodules.

Author: I'll at least add a note to the "Modifying OS images" doc so people are aware of the issue and know what to do.

> git submodules update --init --recursive

Reviewer: ```suggestion
git submodule update --init --recursive
```

- The OS packages build step takes a version string that is used for the new OS image version.

Both scripts will, based on a given version string:
1. check out a respective version tag in both `coreos-overlay` and `portage-stable`
2. update the `version.txt` file accordingly
3. create a new commit with the above changes
4. tag the commit with the version string and push the tag.

## Build automation and build steps

The Flatcar Container Linux build process consists of

1. compiling packages from source, and generating a new OS image release version / tag
2. creating a generic OS image file from the resulting binary packages
3. creating one or more vendor-specific image files from the generic OS image.

Optionally, the build process may include building the SDK from scratch based on a previous - existing - SDK, e.g. to update core build tools and utilities.
In that case, the above 3 steps are preceded by

1. Compile all core and SDK packages from source to generate a new SDK root FS and build a tarball from that; generate a new SDK release version and set the OS release to the same (SDK) version.
2. Build a base SDK container image using the tarball from 1.

>    a. build amd64 and arm64 toolchains and related board support
>    b. then, from the image from 2.a, generate from scratch 3 container images - "all", "amd64", and "arm64" with the respective board support included.

Reviewer 2: Is there somewhere a rendering of this document available? Because apparently github markdown does not support `a.` or `a)`, only `1.` or `1)`. And the nested ordered list uses roman numerals. So:
```suggestion
   1. build amd64 and arm64 toolchains and related board support
   2. then, from the image from 2. i., generate from scratch 3 container images - "all", "amd64", and "arm64" with the respective board support included.
```

Running all 5 steps in one go will produce a new SDK and new OS image based on that new SDK.
In this pipeline, both a new SDK version as well as a new OS image version are generated.
This is what we call a "full" (or "full-all") build.
The main use case is for nightly builds of the "main" branches where development of new features happen.
A new major version release will also use this process (and can be seen as a "special case" of a nightly build of "main").
New major releases always include a new SDK.

Running only the 3 OS image steps is used for active (i.e. supported) release branches.
This uses an existing SDK and thus only generates a new OS image version.
Usually, stabilisation of a major release (alpha -> beta -> stable) uses the same SDK release during its lifetime, so there's no need to always build the SDK.
Only in rare cases it is necessary to update the SDK after a new major version has been published.

### Automation scripts

The [build automation scripts][scripts-repo-ci] reflect the 5 steps outlined above; each step is done in a separate script.
Check out the build automation's README.md to get an overview.

> Each of the scripts contains documentation of the INPUTs and OUTPUTs of the respective build step:

Author: It's all-caps in the scripts' documentation, but I agree to user lowercase here for readability.

The build scripts should run on most Linux-based nodes out of the box; `git` and `docker` are the only requirements.
In the Flatcar project, we use Flatcar Container Linux on our CI worker nodes.

### Auxiliary infrastructure

Apart from infrastructure to run the CI / builds on we also need a server for caching build artifacts.
Build artifacts are mostly container images - with only few exceptions - and are almost always huge (some gigabytes).
To not overly pollute CI workers' disk space, the build scripts support an "artifact cache" server.
Requirements for this server are rather simple - it should have sufficient disk space (we use 7TB on Flatcar's CI and can hold ~50 past builds), ssh access (for rsync) and serve artifacts from the (rsync/ssh) path prefix via HTTPS.
See the `BUILDCACHE_…` settings in the [CI automation settings file](https://github.com/flatcar-linux/scripts/blob/main/ci-automation/ci-config.env) for adapting the build scripts to your environment.

[scripts-repo-ci]: https://github.com/flatcar-linux/scripts/tree/main/ci-automation

> [mod-cl]: sdk-modifying-flatcar.md

Author: ```suggestion
[mod-cl]: sdk-modifying-flatcar
```

## Tips and Tricks

We've compiled a [list of tips and tricks](sdk-tips-and-tricks) that can make working with the SDK a bit easier.

> [scripts-repo-ci]: https://github.com/flatcar-linux/scripts/ci-automation

Reviewer 2: That's 404.

> [scripts-repo-ci]: https://github.com/flatcar-linux/scripts/ci-automation

Author: ```suggestion
https://github.com/flatcar-linux/scripts/tree/main/ci-automation
```
